lib/wca_live_web/controllers/competition_controller.ex

```diff
@@ -174,4 +190,76 @@ defmodule WcaLiveWeb.CompetitionController do
         {:error, "the provided token is not valid"}
     end
   end
+
+  defp fetch_competition_by_id_or_wca_id(id_or_wca_id) do
+    case Integer.parse(id_or_wca_id) do
+      {id, _} -> Competitions.fetch_competition(id)
+      _ -> Competitions.fetch_competition_by_wca_id(id_or_wca_id)
+    end
+  end
+
+  defp format_results(competition) do
+    competition
+    |> Repo.preload(
+      competition_events: [rounds: [:competition_event, results: [:person]]],
+      people: []
+    )
+    |> competition_to_results()
+  end
+
+  defp competition_to_results(competition) do
+    result_registrant_ids =
+      for competition_event <- competition.competition_events,
+          round <- competition_event.rounds,
+          result <- round.results,
+          do: result.person.registrant_id,
+          into: MapSet.new()
+
+    %{
+      events:
+        competition.competition_events
+        |> Enum.map(&format_competition_event/1),
+      persons:
+        competition.people
+        |> Enum.filter(&(&1.registrant_id in result_registrant_ids))
+        |> Enum.map(&format_person/1)
+        |> Enum.filter(&(&1["id"] != nil))
+    }
+  end
+
+  defp format_competition_event(competition_event) do
+    %{
+      "eventId" => competition_event.event_id,
+      "rounds" =>
+        competition_event.rounds
+        |> Enum.map(&format_round/1)
+    }
+  end
+
+  defp format_round(round) do
+    %{
+      "number" => round.number,
+      "results" =>
+        round.results |> Enum.map(&format_result/1) |> Enum.filter(&(&1["attempts"] != []))
```

Comment:
Again, just an optimisation to do a single pass (and filtering upfront is nicer :D).

```suggestion
        for result <- round.results, result.attempts != [] do
          format_result(result)
        end
```